app/client/src/workers/validations.ts

```
@@ -106,105 +114,134 @@ function validateArray(
+    uniqueKeys = compact(
+      allowedKeyConfigs.map((allowedKeyConfig) => {
+        // TODO(abhinav): This is concerning, we now have two ways,
+        // in which we can define unique keys in an array of objects
+        // We need to disable one option.
+
+        // If this key is supposed to be unique across all objects in the value array
+        // We include it in the uniqueKeys list
+        if (allowedKeyConfig.params?.unique) return allowedKeyConfig.name;
+      }),
+    );
+  }
+
+  // Concatenate unique keys from config.params?.unique
+  uniqueKeys = Array.isArray(config.params?.unique)
+    ? uniqueKeys.concat(config.params?.unique as Array<string>)
+    : uniqueKeys;
+
+  // Validation configuration for children
+  const childrenValidationConfig = config.params?.children;
+
+  // Should we validate against disallowed values in the value array?
+  const shouldVerifyAllowedValues =
+    Array.isArray(allowedValues) && !!allowedValues.length;
+
+  // Do we have validation config for array children?
+  const shouldValidateChildren = !!childrenValidationConfig;
 
-    const allowedKeys = (config.params.children.params?.allowedKeys || []).map(
-      (key) => key.name,
+  // Should array values be unique? This should applies only to primitive values in array children
+  // If we have to validate children with their own validation config, this should be false (Needs verification)
+  // If this option is true, shouldArrayValuesHaveUniqueValuesForKeys will become false
+  const shouldArrayHaveUniqueEntries = config.params?.unique === true;
+
+  // Should we validate for unique values for properties in the array entries?
+  const shouldArrayValuesHaveUniqueValuesForKeys =
+    !!uniqueKeys.length && !shouldArrayHaveUniqueEntries;
+
+  // Verify if all values are unique
+  if (shouldArrayHaveUniqueEntries) {
+    // Loop
+    if (
+      uniq(value.map((entry) => JSON.stringify(entry))).length !== value.length
```

If we replace `_.uniq` with our own implementation using loops, we could detect duplicate value and return error early rather than comparing the unique's length with `value.length`.
@riodeuno Do you think we will be gaining significant performance if we do so?

---

Good point @aswathkk 
Not a significant performance gain, but will be notable with large data.
We won't know when JSON.stringify will comeback to bite us, any way to reduce its calls will be good.

If we are not going with our implemetaion, we should use `Set`, it is twice as fast.
https://www.measurethat.net/Benchmarks/Show/3889/0/lodash-uniq-vs-set

---

Sounds good. 
If we go ahead with Set, do you recommend we convert the value into a set at the beginning of the validation and convert it back into an array for use in the widgets, or use Sets only for this particular uniq validation? We have seen data with 50K entries.

---

So, I'm going ahead with Satish's suggestion of using Set instead of _.uniq and stringify. 

@aswathkk I agree that it would be helpful to identify and inform an Appsmith developer regarding the index of the duplicate entry.

I am assuming a standard scenario of a happy path, where the validation passes, in this case the Set approach seems more performant than the loop approach.
Based on this, let's take a call in the interest of performance and circle back when the need to identify and inform arises.

Does that sound good?

---

@riodeuno ,
Even with set approach, it has to go through the whole array in happy path.
It would be nice if we can fail early, if we are not doing it put a note in the code about possible improvement.

---

Yeah, one iteration is inevitable. Given the discussion, we seem to conclude that converting a large array into a Set, would be faster than iterating through a standard array (which is the type of the original input to the validations), in the happy path.

I find getting that performance improvement more valuable than the DX improvement, in this scenario.

I will add a comment.

---

@SatishGandham Just confirmed this. Seems like a major difference in performance.

https://www.measurethat.net/Benchmarks/Show/17083/0/create-set-vs-loop